### ISSUE TITLE
Chocolatey Toast Notifications Extension (acces denied)

### ISSUE THREAD

**Issue by ds2edv** [bcurran3/ChocolateyPackages_CCR#189]
If I user your extension I allways get an acces denied (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED)) error.

Do you have an idea what causes this?

**Comment by bcurran3** [comment]
No.
A quick Google search of that error hints that you may not have administrative access on the computer.
I don't think you need it to use it.
I would remove and reinstall as administrator and see if that helps.
If that doesn't help, I'll need to see logs of the install, specifically if there's anything that says ERROR.

**Comment by ds2edv** [comment]
No I run it as domain admin.

**Comment by ds2edv** [comment]
To clarify - the installation of the extension runs fine. The error is by upgrading, install or uninstall packages as admin.

**Comment by ds2edv** [comment]
I don't use the agent but is completely the same issue like issue # https://github.com/bcurran3/ChocolateyPackages/issues/74

NOTE: EDITED so it cross references. (bcurran3)

**Comment by bcurran3** [comment]
In PowerShell paste in:
`New-BurntToastNotification -text "This is a test."`
and see if you get a toast message.

If not, then a problem with https://github.com/Windos/BurntToast.
If so, permission problem still somehow related to the extension or BurntToast.

Either way, make sure you remove and re-install as previously mentioned.

**Comment by bcurran3** [comment]
So funny thing happened while looking at the code for this:
I found a bug in the 0.0.3 release with the new feature of sending messages to the foreground. I doubt it's related to your problem, but you can try it:
`choco install chocolatey-toast-notifications.extension --version 0.0.3.1`

**Comment by ds2edv** [comment]
I'm a little bussy now. Will give it a try at weekend.

**Comment by ds2edv** [comment]
Sorry, for late reply.

Now it works if you are logged in as user with admin rights and if you run it from rmm as system user.

But still doesn't work if you are logged in as restricted user and run powershell with adminrights and put in admin credeteninals in uac prompt. This always will fails the complete installation of a package. Usecase is from time tome you have a remote session to assit a user and you would install a package without logout the current user but this is impossible if the extion is installed because package installation fails.

Do you see a solution for this? Maybe don't show the toast in this circumstances?

Maybe there is a way to combine your scrit with the RunAsUserModule (https://github.com/KelvinTegelaar/RunAsUser)? I use it often to show toasts to logged in user in other scripts.

**Comment by bcurran3** [comment]
Thanks. Forward progress. Good.

I don't think this will be hard to fix to work under the "restricted user and run powershell with adminrights and put in admin credeteninals in uac prompt" scenario. I just need to understand the conditions in that scenario so I can put in a check for some unique condition and change behavior based on it. Yes, probably just skip notifications in that case.

I'll do some testing and see...

**Comment by ds2edv** [comment]
Do youe have anything new about it? I have tested something like below but doesn't work reliable.

# Identify user-situation
$ExplorerProcess = get-process | Where-Object {$_.ProcessName -eq "explorer"}
$ExplorerProcess = $ExplorerProcess.ProcessName
$exp = (Get-CimInstance Win32_Process -Filter "name = 'explorer.exe'")
$expcou = (Get-CimInstance Win32_Process -Filter "name = 'explorer.exe'").count
$pow = (Get-CimInstance Win32_Process -Filter "name = 'powershell.exe'")
$powcou = (Get-CimInstance Win32_Process -Filter "name = 'powershell.exe'").count
$expown =  Invoke-CimMethod -ErrorAction SilentlyContinue -InputObject $exp -MethodName GetOwner | Select-Object user 
$powown =  Invoke-CimMethod -ErrorAction SilentlyContinue -InputObject $pow -MethodName GetOwner | Select-Object user 

# Show toast notification
if ($env:USERNAME -eq "$env:COMPUTERNAME$" -and $ExplorerProcess -eq "explorer") {
  Write-Host "Choco run as sytem and a user is connected - sending toast"
  start-process powershell -ArgumentList "$PSScriptRoot\notify_ins.ps1"}
elseif ($env:USERNAME -ne "$env:COMPUTERNAME$"-and $expown -eq $powown) {
  Write-Host "Admin-User is logged in - sending toast directly"
  New-BurntToastNotification -Text "DS2 - Update-Manager", "$chocolateySoftwareName wurde installiert", "oder auf die Version $env:chocolateyPackageVersion aktualisiert." -AppLogo $chocolateyIcon -Sound 'Reminder'}
elseif ($env:USERNAME -eq "$env:COMPUTERNAME$" -and $ExplorerProcess -ne "explorer") {
  Write-Host "Choco run as system but no user connected - skip toast"} 
elseif ($env:USERNAME -ne "$env:COMPUTERNAME$" -and $expcou -ne 1 -or $env:USERNAME -ne "$env:COMPUTERNAME$" -and$powcou -ne 1) {
  Write-host "Unclear situation about logged in user or you are running choco from uac-promted powershell - skip toast"
 }
else {
  Write-host "Something went wrong - skip toast"}

**Comment by bcurran3** [comment]
I haven't done any testing yet.
I plan to test on my son's machine which has a similar setup when I have a chance.
(Need a time when I'm not busy doing whatever and he's not playing Robolox!)

p.s. 0.8 of BurntToast came out. I have no idea if it changes anything related to this, but doubt it.

p.s.s. The root cause of problems before was that WinRM wasn't running. I checked for it wrong (typo) and that's what the recent fix was. (Supposed to not send the Toast if WinRM is not running under system user.) Make sure WinRM is running on your problem machine.

**Comment by ds2edv** [comment]
Still the same with your mentions.

**Comment by bcurran3** [comment]
So I tested this last night and reproduced the problem.

Logged in as my son who doesn't have administrative privileges, I went to command prompt as admin  (after typing in credentials of course) and did a whoami. It showed me as domain\admin as opposed to domain\mysonsname. Also confirmed via Task Manager that it only showed my son logged in.

What's happening is that the toast is trying to be sent to the user with the admin credentials logged into the CLI instead of the user who's logged into the computer. It can't send to other admin because the admin isn't technically logged in.

This was the same problem when logged in as SYSTEM (e.g. choco-upgrade-all-at scheduled task) and in that case I use WinRM to send the Toast globally instead of the logged in user. 

So basically the fix, and I kinda like this, is to always send the Toast globally.

Easy fix to implement. Means deleting a bunch of code instead of adding to it. :) The caveat is that WinRM will be a hard requirement for it to work. This is a lot easier than checking for logged in users and sending to them (which I might still play with to avoid having to rely on WinRM.)

**Comment by bcurran3** [comment]
I just pushed a pre-release for you to test: chocolatey-toast-notifications.extension.0.0.4-pre

`cup chocolatey-toast-notifications.extension --version 0.0.4-pre`

This should take care of the problem you have as well as take care of https://github.com/bcurran3/ChocolateyPackages/issues/74 once and for all.

Please test and let me know.

There will be changes before it's released. I'm going to check for WinRM running and if not, run it, send toast, and then stop WinRM again.

**Comment by ds2edv** [comment]
Hi seems to be working.

**Comment by bcurran3** [comment]
Great.
I just pushed 0.0.4 official. It'll probably be a few days before it's through moderation. Give this final a whirl if you have the time.

https://chocolatey.org/packages/chocolatey-toast-notifications.extension/0.0.4

I'm going to close this out.
